tools/ResourceUpdater/main.cpp

```diff
@@ -795,11 +813,29 @@ bool update_battle_chars_info(const std::filesystem::path& input_dir, const std:
 
     auto& chars = result["chars"];
     std::map<std::string, std::vector<std::string>> tokens;
-    for (auto& [id, char_data] : chars_json.as_object()) {
+    for (auto& [id, char_data] : chars_cn_json.as_object()) {
         json::value char_new_data;
         std::string name = char_data["name"].as_string();
-
         char_new_data["name"] = name;
+
+        if (chars_en_json.find(id)) {
+            std::string name_en = chars_en_json.as_object()[id]["name"].as_string();
```

Comment:
```c++
        if (auto char_en = chars_en_json.find(id)) {
            char_new_data["name_en"] = char_en["name"];
        }
```

Reply:
If not found, do we need to set `char_new_data["name_en"]` to null or id or something?

Reply:
If there is no corresponding "global" Id we could just use the CN name. But in EN case, we have the official translations for the operators, the id should also remain the same the problem becomes populating the jsoj for this case.
Or... (This requires more work) check what client is selected and with that we know how many operators there are in that server.
YostarEN has less ops than Official CN

Reply:
> ```c++
>         if (auto char_en_opt = chars_en_json.find(id)) {
>             char_new_data["name_en"] = char_en_opt->at("name");
>         }
> ```

Is std::string name needed for the cn? Or we can just remove that variable like you are suggesting for the overseas?

Reply:
> If not found, do we need to set `char_new_data["name_en"]` to null or id or something?

I'm having a look at this at this moment. We could go two ways:
- ResourceUpdater can give the overseas names the CN name if the ID is missing (this way in the UI we should not have anything to worry about)
- or we can have an empty string (ex: `"name_en" = ""`,) and then check in the UI if the names are empty to use the CN name. At the end of the day, the result is the exact same. It will show the CN characater.

Technically, EN has all the names CN has. The problem becomes comparing the two resource jsons, because the EN doesn't have the IDs so it will always give a false.

Reply:
> > ```c++
> >         if (auto char_en_opt = chars_en_json.find(id)) {
> >             char_new_data["name_en"] = char_en_opt->at("name");
> >         }
> > ```
> 
> Is std::string name needed for the cn? Or we can just remove that variable like you are suggesting for the overseas?

OK, we can remove `as_string`

Reply:
Yep, did that yesterday.

Reply:
An empty string is probably not a good idea, it means there is a value here (albeit an empty value), not that there is no value. I think

- The callback given by core does not contain name_en. If the UI checks that there is no name_en field, the name field is used.
- The callback given by core always contains name_en, but it may be Chinese, and the UI uses it directly.

Either way is possible, it’s up to you.

Reply:
Is there the "best performance solution"? Because I've noticed that it's a pretty demanding task